List View: Add multi-select behaviour for blocks when shift key is selected

andrewserong opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Fixes: #37514 

This PR borrows logic from the `useMultiSelection` hook to implement multiple block selection from the list view within the post, site, and widget editors, and the list view in the Navigation block. The goal is that wherever the `ListView` component is used, it comes with multi-block selection via the SHIFT key by default.

With the SHIFT key held down, you should be able to select multiple blocks from the same nesting level.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

### Test shift+click behaviour

1. In the post editor, try selecting multiple blocks from the list view and re-arrange the blocks by dragging.
2. Running a blocks-based theme, In the site editor, try selecting multiple blocks from the list view and re-arrange the blocks by dragging. Also, try selecting a few blocks together and click on the contextual menu in the editor canvas and group the blocks.
3. Running a classic theme, open up the widgets editor, and try selecting multiple blocks from the list view and re-arrange the blocks by dragging.
4. In a post, page, or site editor view, add a Navigation block with several child blocks. With the parent Navigation block selected, select the block toolbar option to view the list view (see screengrab below) and shift-click to select multiple blocks and then drag to rearrange.

### Test shift+up/down arrow behaviour

1. Try tabbing to the list view and shift+press up and down arrows to select multiple blocks. Note that shift+selecting from within a selection should extend the selection from where the selection began. If you shift+select from a block that is not adjacent to the current selection, it should start a new selection.

There are an additional two e2e tests added in this PR to cover mouse and keyboard behaviour. You can run them locally via:

```
npm run test-e2e -- --testPathPattern packages/e2e-tests/specs/editor/various/multi-block-selection.test.js
```

While running locally, I noticed that some of the other tests failed intermittently. If you wish to run this locally, too, to make it easier for local testing, you can update the `it(` lines in the `multi-block-selection.test.js` file to `it.only(` to only test the added tests.

There is also an added unit test which can be run via:

```
npm run test-unit -- --testPathPattern packages/block-editor/src/components/list-view
```

## Screenshots <!-- if applicable -->

| Post editor | Site editor |
| --- | --- |
| ![Kapture 2022-02-01 at 16 16 13](https://user-images.githubusercontent.com/14988353/151916387-204f1c19-87d0-4b7c-8766-b3357ca8f50b.gif) | ![Kapture 2022-02-01 at 16 19 42](https://user-images.githubusercontent.com/14988353/151916691-3f475d76-8bf8-4e59-866e-f174d398971e.gif) |
| Widget editor | Navigation block list view |
| ![Kapture 2022-02-01 at 16 08 33](https://user-images.githubusercontent.com/14988353/151915796-7bdd8aa1-be70-4763-990b-d8bfc64508ec.gif) | ![Kapture 2022-02-01 at 15 54 46](https://user-images.githubusercontent.com/14988353/151914871-5d549bb8-9e5a-4992-ab83-f714534c8e8c.gif) |

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
New feature (non-breaking change which adds functionality)

## To-do

* [x] If no blocks are selected and you go to shift-select the first block, it should select it
* [x] Multi-select behaviour + drag and drop to rearrange blocks doesn't appear to be working properly (it only moves the first item instead of all selected items, and the selection after rearranging the blocks doesn't appear to be quite right)
* [x] There's duplication between post the `edit-post` and `edit-site` behaviours — can we move the select logic to a hook to remove the duplication? Update: yes! I've moved the logic to the component since we'd like this logic in place in each of the use cases for the list view.
* [x] Add tests

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [ ] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

andrewserong committed:
List View: Add multi-select behaviour when shift key is selected

andrewserong committed:
Ensure shift clicking a block when no blocks are selected selects that first block

andrewserong committed:
Add support for dragging multiple selected blocks

andrewserong commented on packages/block-editor/src/components/list-view/block-contents.js:
```diff
@@ -32,17 +32,21 @@ const ListViewBlockContents = forwardRef(
+					getSelectedBlockClientIds,
 				} = select( blockEditorStore );
 				return {
-					rootClientId: getBlockRootClientId( clientId ) || '',
```
As of https://github.com/WordPress/gutenberg/pull/33320 the `rootClientId` is no longer used in this function, so I think it can safely be removed here.

andrewserong committed:
Remove duplication by moving multi-select behaviour to the ListView component, ensure it also works in the widget editor

andrewserong committed:
Add inline comments

andrewserong committed:
Update documentation, add changelog entry

andrewserong committed:
Add e2e test for multi-select in the list view

andrewserong commented:
Hi folks! (Looks like this PR touches files that ping quite a number of folks!) 👋

I don't have a lot of experience making changes to the `ListView` component, so I'm keen for feedback on whether or not it looks okay moving the multi select logic to the component. It seems to work pretty well for me in testing (across the post, site, and widget editors, and the navigation block's list view), but let me know if you run into any issues!

noisysocks commented:
Wow this works insanely well! Nice job.

I tested: selecting multiple blocks and using keyboard shortcuts (e.g. <kbd>Del</kbd>, <kbd>Cmd</kbd>+<kbd>c</kbd>), selecting multiple blocks and dragging them within the same parent, selecting multiple blocks and dragging them into a different parent, selecting multiple blocks that have children, and doing this across each of the four editors. (Thank you for remembering that we have multiple editors!)

I set Firefox to emulate an iPad in landscape mode and noticed that I could not select multiple blocks. (iPads support bluetooth keyboards.) But you can't do this using the block editor either, so definitely not a blocker for this PR.

My only thought is that maybe we should hide the ⋮ buttons when multiple blocks are selected? They're a bit confusing in this context. It's not clear if clicking on one and selecting an action will perform that action on one or multiple blocks.

<img width="363" alt="Screen Shot 2022-02-02 at 16 52 33" src="https://user-images.githubusercontent.com/612155/152100398-2fd31495-e180-4f1e-9c07-bbcd1f2f027b.png">

noisysocks commented on packages/block-editor/CHANGELOG.md:
```diff
@@ -6,6 +6,11 @@
+### New Features
+
+-   List View now supports selecting and dragging multiple blocks via `SHIFT` clicking items in the list [#38314](https://github.com/WordPress/gutenberg/pull/38314).
+>>>>>>> e0937d166f (Update documentation, add changelog entry)
```
```suggestion
```

andrewserong replied:
Thanks! That wound up in there from a bad rebase 😀

noisysocks commented on packages/block-editor/src/components/list-view/index.js:
```diff
@@ -89,11 +98,64 @@ function ListView(
 		[ draggedClientIds ]
 	);
 	const selectEditorBlock = useCallback(
```
This callback does a lot more than selecting an editor block now so maybe its name should change to `updateBlockSelection` or something like that.

andrewserong replied:
Oh, nice idea. Happy to update it 👍

noisysocks commented on packages/block-editor/src/components/list-view/index.js:
```diff
@@ -89,11 +98,64 @@ function ListView(
+					}
+				}
+			}
 			onSelect( clientId );
```
Should we pass multiple IDs to this prop?

Or, since it looks like we're no longer using it, remove the prop altogether? Maybe third parties would like to have it though.

`ListView` is currently experimental so we can safely make changes to its API.

andrewserong replied:
I actually really like the idea of removing it altogether if we can. Happy for any feedback folks might have!

andrewserong committed:
Remove stray line from changelog

Co-authored-by: Robert Anderson <robert@noisysocks.com>

noisysocks commented:
> I'm keen for feedback on whether or not it looks okay moving the multi select logic to the component.

@gwwar and @talldan would know better but I think it seems pretty reasonable. It's interesting that `ListView` was already (unnecessarily?) calling `selectBlock`.

andrewserong commented:
Thanks for reviewing and taking this for a spin @noisysocks! 🙇

> My only thought is that maybe we should hide the ⋮ buttons when multiple blocks are selected? They're a bit confusing in this context. It's not clear if clicking on one and selecting an action will perform that action on one or multiple blocks.

That's a good question — currently the ⋮ button adjusts settings for the individual block. Personally, I'd lean toward making changes to the UI in subsequent PRs if we can. Another nice thing to look into would be how we visualise dragging multiple blocks so that it's clear it isn't the one block that's being dragged. But my thinking was, if we can land the multi-select first, then it'll hopefully be easier to look at some of these more granular UI things in isolation.

Happy to do it here, though, if folks think it's a blocker!

talldan commented:
Wow, this is a pretty cool feature! Appreciate all the hard work.

> @gwwar and @talldan would know better but I think it seems pretty reasonable. It's interesting that ListView was already (unnecessarily?) calling selectBlock.

I don't have any other suggestions, so this seems good. Might be nice to wrap it up into a hook or something.

It'd be good to make sure this gets some accessibility testing. @alexstine has been working on List View a lot recently so can possibly assist. 

One thing that jumps out is that some spoken messages might be required to announce the selection range. Though potentially that could already be implemented in the multiselect store actions. Worth double-checking that and whether any existing messages are still up to the task.

(edit: I notice now this is only usable via shift+click, it should probably work via arrow keys too. It could be done in stages via a separate PR, but without that it's not really an accessible feature)

ramonjd commented:
Here's what I'm seeing

![Feb-02-2022 20-42-49](https://user-images.githubusercontent.com/6458278/152129719-5223f2ab-11d0-4ce1-bb09-54d3ee9d4da1.gif)


I tested with

👍  `SHIFT` clicking and dragging/dropping collapsed and non-collapsed blocks. 

👍  `SHIFT` clicking blocks in the editor: the corresponding blocks are highlighted in the list

👍  `SHIFT` clicking expanded and top level list items: the nested blocks aren't displayed as selected, which is what I'd expect

<img width="100" alt="Screen Shot 2022-02-02 at 8 39 42 pm" src="https://user-images.githubusercontent.com/6458278/152129189-b56f59dc-db98-48b1-ac6d-6dde1649e50b.png">

👍  `SHIFT` clicking and selected 30+ blocks at a time. Perceived performance was the same as dragging one.

This is a wonderful UX improvement, thank you! And for the e2e tests.

> Might be nice to wrap it up into a hook or something

+1

gwwar commented:
> It's interesting that ListView was already (unnecessarily?) calling selectBlock.

🤔 I don't remember offhand, but I'm still catching up on things after my 3 month sabbatical too. 😆 I'll try and 🔍 this one today.

This is a bit of a tangent, but one other thought that comes to mind is that I'd really like to optionally run performance tests with list view open when we make changes in this component. I'll maybe smoke test in a large post for now, and think through some options in another PR.

gwwar commented:
Reserving this comment for manual testing:

TIL that <kbd>shift</kbd> + mousewheel = scroll horizontally

✅ In a large post (where we end up using windowing), the selection behaves as expected

https://user-images.githubusercontent.com/1270189/152197605-6e868f53-e034-473d-ae60-d65be5f86bae.mp4

❌  This might be arguable/better left for a follow up PR. A selection starting from nested child blocks, ends up selecting the parent. If you compare behavior in a normal filesystem, this isn't the case.

https://user-images.githubusercontent.com/1270189/152198122-fb7b5598-3790-4565-8874-709151c7ac79.mp4

🤔 Looking at some older videos, we didn't use to select blocks in the modal view (behavior is now in trunk too). Was this an intentional change?

❌ Open List View, also open the navigation modal. Nav menu is removed and inserted to bottom of post:

https://user-images.githubusercontent.com/1270189/152202033-d004409d-9d44-4822-8af4-62c7bc5186e2.mp4

gwwar commented on packages/block-editor/src/components/list-view/index.js:
```diff
@@ -89,11 +98,64 @@ function ListView(
+					const depth =
+						Math.min( startPath.length, endPath.length ) - 1;
+					const start = startPath[ depth ];
+					const end = endPath[ depth ];
```
🤔 It may be worth decomposing this a bit and unit testing the calculation of start and end. We probably need to document a few corner cases with selections across nested children.

gwwar commented on packages/block-editor/src/components/list-view/block-contents.js:
```diff
@@ -56,7 +60,7 @@ const ListViewBlockContents = forwardRef(
 
 		return (
-			<BlockDraggable clientIds={ [ block.clientId ] }>
+			<BlockDraggable clientIds={ selectedBlocks }>
```
I don't think we can make this assumption in all cases. It's possible to start dragging a target in the list view sidebar/modal without selecting the block. (Block selection is mousedown+mouseup, while a drag start is just mousedown).

https://user-images.githubusercontent.com/1270189/152205347-f7b20507-9e99-43bf-bf08-94bdfdb7e38f.mp4

andrewserong replied:
Great catch @gwwar, and thanks for all your detailed testing here. If I don't get to addressing all these issues today, I'll continue on working through them next week! 🙇

andrewserong replied:
I think I've come up with a fix for this. In https://github.com/WordPress/gutenberg/pull/38314/commits/91997f461e4b5492466108cc953f603c42eb1071 I've updated it so that the list of selectedBlocks is only passed in if the clicked on block is part of that selection, otherwise it uses the client id of the clicked on block. This appears to restore the behaviour where if no block is selected, dragging still works. It also appears to work correctly if you have a few blocks selected and then go to drag a block outside of that selection, but it probably needs a bit more manual testing to confirm 🙂

gwwar commented:
Overall this is coming along quite nicely @andrewserong ✨ 

A few things we'll want to clarify:

- I think the duplicated selection code in the sidebar was because we didn't intend for blocks to be selectable in other list views, like the nav modal. I think it's okay to change the behavior, provided that this is an intentional change.
- Unit tests for the start/end selection calculation will be really useful for thinking through complex selection cases https://github.com/WordPress/gutenberg/pull/38314#discussion_r797835922
- We'll want to fix drag targets as noted in https://github.com/WordPress/gutenberg/pull/38314#discussion_r797848377 
- Some other things I spotted in https://github.com/WordPress/gutenberg/pull/38314#issuecomment-1028135117 can maybe be left for follow up PRs
- +1 for getting a11y feedback early

alexstine commented:
@andrewserong Can you tell me if I am using this right?
1. Open the List View.
2. Find a couple Blocks to select.
3. Hold Shift and press Down Arrow to select my Blocks.

My problem is I can't hear any announcement of what has been selected. Is it possible to fire a wordpress/a11y.speak() on each selection/deselection?

```
speak( 'Paragraph selected' );
speak( 'Paragraph unselected' );
```

Thanks.

andrewserong commented:
Thanks for testing @alexstine, much appreciated! This PR currently only focuses on adding in mouse behaviour for multiple block selection while holding down the shift key. Adding in the keyboard based behaviour and a call to `wordpress/a11y.speak()` is a great idea, though, thank you, we should get that covered. I'll look into it!

andrewserong committed:
Ensure that clicked on blocks that aren't a part of a selection can still be repositioned

andrewserong committed:
Try a naive approach to keyboard handling for shift + up/down to select multiple blocks

andrewserong commented:
Thanks again for the feedback and testing everyone! I'm finishing up for the week, but will continue on with this PR on Monday.

Before splitting out the `selectEditorBlock` function to a separate hook and adding further tests, I thought I'd have a go at adding in keyboard handling. In [this commit](https://github.com/WordPress/gutenberg/pull/38314/commits/eed3709c91f904fc17c65ee13528a5ab6028e165) I've tried a naive attempt at allowing shift plus the up or down key to fire the onClick event to select the currently highlighted block. This then defers the shift plus up or down key behaviour to the editor canvas where multiple block selection is already handled. This seemed a little simpler than trying to duplicate the behaviour within the list view itself. However, I only just got that technically working before finishing up for the day, so I haven't tested it properly yet (and I was also a little concerned it might be hacky in its approach). Please let me know if anyone thinks it needs its own handling, instead, of course!

On Monday, I'll take another look at the keyboard handling and give it some more testing, then will look into splitting the select handler out to a hook and adding some unit tests for the logic there 🙂

andrewserong committed:
Move block selection to its own useBlockSelection hook

andrewserong committed:
Refactor start and end id calculation to its own function, add unit tests

andrewserong commented:
I've refactored to a hook in https://github.com/WordPress/gutenberg/pull/38314/commits/ed38d5cd5818a91cddeba888a0526b6e5068ffe1 and rearranged a tiny bit of the start/end id logic to a separate function with unit tests in https://github.com/WordPress/gutenberg/pull/38314/commits/47e85b05d629e0b91ab1ee397d98b5febb34cfcd — it still needs some inline documentation and possibly tidying up a little bit more, but I _think_ it's going in the right direction. Will continue on with this more tomorrow afternoon.

andrewserong committed:
Move utility function to utils file, add doc comment, update usage

andrewserong commented:
@gwwar if you get a chance, I think this PR is now ready for another review. I've moved the block selection logic to a hook, added a unit test, and tidied up some of the dragging behaviour. Personally, I like that the Navigation modal list view allows blocks to be selected (I think it might make the modal slightly more useful that just being used for rearranging blocks), but let me know if folks feel differently!

> This might be arguable/better left for a follow up PR. A selection starting from nested child blocks, ends up selecting the parent. If you compare behavior in a normal filesystem, this isn't the case.

I borrowed the nesting logic there from the other multiple block selection logic we have elsewhere. For the moment at least, I think it's probably worth preserving for the safety of rearranging blocks, but it'd be a good thing to play with in potential follow-ups. Hopefully the added unit test / split out utility function helps document the behaviour a little.

andrewserong commented:
@alexstine this is also ready for more testing of the keyboard behaviour. It's technically working for me, but I'd love feedback whenever you get the chance, on whether or not deferring to the editor canvas for the multiple block selection is viable and accessible (or how it can be improved). Thanks!

alexstine commented:
@andrewserong Whenever I start selecting in the List View by holding down Shift+Down Arrow, focus is shifted to the actual Block area as I select Blocks. Whenever I am done selecting, I am a long ways down the DOM from the List View itself. Is this what is supposed to happen? These UX don't really work if so.

gwwar commented:
@andrewserong I verified that the dragging behavior feels much better. ✨  Thanks for also adding additional unit tests.

> Whenever I start selecting in the List View by holding down Shift+Down Arrow, focus is shifted to the actual Block area as I select Blocks. 

Great catch @alexstine! 

The primary action when selecting a list item is focusing the block in the main editor. I'm not a huge fan of this behavior (see https://github.com/WordPress/gutenberg/issues/29466 ), but we need to avoid the primary action when <kbd>shift</kbd> is held, otherwise this feature is going to be for mouse users only. Getting focus back to the list view when this happens can be quite an ordeal. 

If folks haven't already, I'd highly recommend for y'all to use a screenreader to see how disorienting it is. Voiceover is available out of the box in osx. Note that some system settings may need to be modified for proper testing https://www.a11yproject.com/posts/2017-12-29-macos-browser-keyboard-navigation/

Here's a video of what this looks like when we use a keyboard and lose ListView focus when selecting:

https://user-images.githubusercontent.com/1270189/153033928-01826a4f-8b3b-491b-94d0-fc15c5d93b2f.mp4

Similarly the modal view also has the wrong focus when selecting multiple blocks. I've added a short video below. If we do manage to get a proper selection using <kbd>shift</kbd> and arrows, what's the expected shortcut for repositioning items using a keyboard only?

https://user-images.githubusercontent.com/1270189/153035882-439514a5-3049-4e58-b58f-da40225e0bd4.mp4

gwwar commented:
As a quick follow up note, the modal view is likely going to be removed in favor of a new type of sidebar area. So I'd avoid too much extra effort if we're having trouble in that modal view specifically. https://github.com/WordPress/gutenberg/issues/36667

andrewserong commented:
That's fantastic feedback @alexstine and @gwwar, thanks again for collaborating with me on this one! I'll dig in and see how we might keep the focus within the list view while shift selecting multiple blocks. Cheers!

andrewserong committed:
Update multiSelect behavior to support keeping focus within the ListView while selecting multiple blocks

andrewserong committed:
Update unit tests, add additional e2e test for keyboard behaviour

andrewserong committed:
Revert change to focus for when shift key is not held

talldan commented on packages/block-editor/src/components/list-view/index.js:
```diff
@@ -57,7 +57,6 @@ const expanded = ( state, action ) => {
 function ListView(
 	{
 		blocks,
-		onSelect = noop,
```
I noticed this prop is in the code example in the README file for List View, so it can be removed from there too.

andrewserong replied:
Removed.

andrewserong commented:
I've updated this PR to now retain the focus within the list view while shift-selecting blocks in the list. It borrows a bit of the `initialPosition` logic from `selectBlock` and includes it in `multiSelect`.

This also implements a particular approach for keyboard navigation with the shift key held down. In this approach, if you start a selection with the shift key held, then pressing up and down will expand that selection. If you then move to a block that is not adjacent to the current selection and shift-select again, it will start a new selection.

I'm not sure if this is the correct behaviour, but while playing around with the interface, this felt like the best trade-off I could manage so that it's possible to start a new selection via keyboard. And it does feel much better now keeping the selection within the list view.

It could do with some more manual testing when anyone has the chance, and I'm very happy for feedback on how to improve it further! And there are probably some things (like re-arranging selected blocks by keyboard) that might be better explored in follow-ups now that the PR is getting fairly big. Thanks again 🙇

talldan commented on packages/block-editor/src/components/list-view/use-block-selection.js:
```diff
@@ -0,0 +1,129 @@
+			if ( ! event.shiftKey ) {
+				await clearSelectedBlock();
+				selectBlock( clientId );
+			} else if ( event.shiftKey ) {
```
This might be a good candidate for an early return to reduce nesting, e.g.:

```jsx
if ( ! event.shiftKey ) {
    // Select a single block.
    await clearSelectedBlock();
    selectBlock( clientId );
    return;
}

// Select multiple blocks.
// ...

andrewserong replied:
Yes, that's a fair bit neater. I've switched to an early return.

talldan commented on packages/block-editor/src/components/list-view/use-block-selection.js:
```diff
@@ -0,0 +1,129 @@
+					)
+				) {
+					// Ensure that shift-selecting blocks via the keyboard only
+					// expands the current selection if hovering over already
```
```suggestion
					// expands the current selection if focusing an already
```

andrewserong replied:
Thanks! Yes, these should all be focused/focusing instead. I've updated the comments in that file 👍

talldan commented on packages/block-editor/src/components/list-view/use-block-selection.js:
```diff
@@ -0,0 +1,129 @@
+				) {
+					selectBlock( clientId, null );
+					return;
+				}
```
It might be possible to move this above the previous `if` statement as a small optimization (when handling a click we wouldn't have to run keyboard related logic).

andrewserong replied:
Good catch, I've moved it up.

talldan commented on packages/block-editor/src/components/list-view/use-block-selection.js:
```diff
@@ -0,0 +1,129 @@
+					return;
+				}
+
+				// Handle keyboard behavior for selecting multiple blocks.
```
The keyboard behavior gets a bit complicated when focus moves into a nested block.

Here's a small example video where the first block is selected and focus, and I press SHIFT+DOWN. Multi-select selects the Query Loop (the next block at the same level), but focus moves to the Group block:

https://user-images.githubusercontent.com/677833/153544692-4d7068fd-32ce-4ae5-9c4e-3a4ab6c124e4.mp4

I think generally my expectation would be that multi-selection follows where my focus is or vice versa. So if I press SHIFT+DOWN, focus moves to the same place that selection does.

There's another case where I try to multi select inner blocks. In the main block list when I try this, my focus is constrained to only blocks at the same level, and I think this might also be the best way for it to work in List View.

I think the problem is going to be that the `onKeyDownHandler` in the BlockSelectionButton isn't adequate for the task, because it's hard to make assumptions about where focus is and secondly it's also difficult or impossible to affect where focus moves to.

There are probably a couple of options:
- Multiselection listens to `onFocus` events that bubble up to the wrapper element of ListView. When focus moves to an adjacent block and shift is pressed, the block selection is expanded. This makes it possible to have multi-selection work in synchronicity with focus, but still makes it difficult to constrain focus movement when selecting inner blocks.
- The TreeGrid component handles multi selection. This has the advantage that focus can be controlled or constrained completely, but it's hard to reason about which block is selected, because TreeGrid has no concept of blocks.

andrewserong replied:
> I think generally my expectation would be that multi-selection follows where my focus is or vice versa

Great catch @talldan, thanks for the detailed look here, and for the suggestions. I tried a couple of things out and settled on deferring how we get the next/prev block id to the `TreeGrid` component by adding in an additional callback `onChangeRow` — this passes up the event object, the originally focused row, and the target focused row, so that we can then use them in `updateBlockSelection`. It also allows us to slightly simplify the logic in the `use-block-selection` hook.

Since we already pass row data in a similar way in `onExpandRow`, it seems to have a pretty good parallel there, while not introducing additional complexity to the `TreeGrid` component directly. Here's how it's looking now (as we press DOWN, the selection and focus moves to the child blocks of Navigation instead of to the next Group block):

![Kapture 2022-02-14 at 17 31 42](https://user-images.githubusercontent.com/14988353/153812508-6fac01ac-c789-493a-b794-8316b6124482.gif)

> In the main block list when I try this, my focus is constrained to only blocks at the same level, and I think this might also be the best way for it to work in List View.

Good point about the constrained focus in the editor 🤔. I'm not sure I agree that the same UX should apply in the list view, though. Because of the free and visible navigation within the tree view, I don't think holding shift should prevent the movement of which element we're focusing. My preference would be for us to keep the click-behaviour multi-select logic so that moving up a level automatically expands the selection to be more broad. E.g. from a deeply nested part of the hierarchy, I think it should be possible to press SHIFT+UP to select a broader region. But this could just be a personal preference 🙂!

talldan replied:
Sounds great, the gif looks really good.

Thanks for working on that solution. I'll test it out a bit later.

talldan replied:
Had a chance to test it now, and it's really good, great work! I'll leave some general comments on the main discussion.

talldan commented on packages/block-editor/src/store/actions.js:
```diff
@@ -271,7 +275,7 @@ export const multiSelect = ( start, end ) => ( { select, dispatch } ) => {
 	}
 
-	dispatch( { type: 'MULTI_SELECT', start, end } );
+	dispatch( { type: 'MULTI_SELECT', start, end, initialPosition } );
```
Probably good to get some wider input from @ellatrix on this additional parameter. And it might also be an idea to make it experimental to begin.

andrewserong replied:
Good idea to get wider input, thanks!

Because this hooks into the existing behaviour of the `initialPosition` state, we probably can't rename it / add an `__experimental` prefix. In this case, we're using the existing logic from initialPosition as used on an individual block in `selectBlock` and re-using the same logic for `multiSelect`, so it's more that the usage of it is experimental rather than that the state is experimental. Let me know if you can think of a good way for us to flag the usage, though!

talldan replied:
I'm mostly hesitant because I don't know `initialPosition` very well.

The `multiSelect` function signature is the public API here really, while the dispatch is an internal API. So I left my comment on the wrong line 😄 

If we want to make it experimental it should be ok to change line 266 to this:
```js
export const multiSelect = ( start, end, __experimentalInitialPosition = 0 ) => ( {
```

and then this line to:
```js
dispatch( { type: 'MULTI_SELECT', start, end, initialPosition: __experimentalInitialPosition } );
```

This way the `multiSelect` function's param is documented as experimental.

andrewserong replied:
Ah, perfect, thanks @talldan of course, that makes good sense now. I'll update the function signature 😀

andrewserong replied:
Updated in https://github.com/WordPress/gutenberg/pull/38314/commits/db566abd56fb685edeabff48f3cbe437ece1db7b

andrewserong committed:
Defer calculation of the next/prev clientId to the TreeGrid component where we can pass the next row that is focused

andrewserong commented:
Thanks for the detailed feedback @talldan! 🙇 I think I've fixed up the main issue there (working out which clientId we want to deal with for the multi-select so it matches the items available in the tree). If the approach looks viable (adding in a `onChangeRow` callback), I'll update the `TreeGrid` component's readme with docs for that and the two other callbacks.

javierarce commented:
Thanks for this PR, great job!

I think I've discovered a small bug that hasn't been mentioned before: If I select a group of items and then shift+click the chevron to expand or collapse an item that contains several elements inside, the group is expanded or collapsed, but the browser opens a new window with the current URL.

I think it should be possible to toggle items with our without the shift key, and that shouldn't affect the selection.

talldan commented on packages/block-editor/src/components/list-view/block.js:
```diff
@@ -104,8 +104,7 @@ function ListViewBlock( {
 		( event ) => {
-			event.stopPropagation();
-			selectBlock( clientId );
+			selectBlock( event, clientId );
```
What's the background behind this change?

Just an FYI, there was a bug that someone noticed yesterday that's related to `selectEditorBlock` - https://github.com/WordPress/gutenberg/pull/38760.

andrewserong replied:
Thanks for the heads-up about that PR! I'll rebase this PR tomorrow and do some extra testing to see if I need to refactor a little more there to preserve that fix. One of the main changes in this PR is that we need to pass up the click and keyboard events to where we use the `updateBlockSelection` function, so that the `useBlockSelection` hook can factor in the type of event, whether the shift key was held, and whether it was a click or keyboard event.

andrewserong replied:
I've adjusted the logic slightly (added in a separate `selectDuplicatedBlock` callback so that `selectBlock` is called correctly), but the behaviour from that fix is now preserved.

talldan commented on packages/components/src/tree-grid/index.js:
```diff
@@ -45,18 +45,26 @@ function getRowFocusables( rowElement ) {
+		children,
+		onExpandRow = () => {},
+		onCollapseRow = () => {},
+		onChangeRow = () => {},
```
It'd be good to update the changelog for this new prop.

andrewserong replied:
Added in https://github.com/WordPress/gutenberg/pull/38314/commits/4c7e257ca34e19950089549d6d4539f1a299691a, but there's a merge conflict that I'll resolve when I rebase this PR tomorrow

talldan commented on packages/e2e-tests/specs/editor/various/multi-block-selection.test.js:
```diff
@@ -747,4 +748,87 @@ describe( 'Multi-block selection', () => {
 		expect( selectedText ).toEqual( 'Post title' );
 	} );
+
+	it( 'should multi-select in the ListView component with shift + click', async () => {
```
There is also a list-view.test.js, not sure if the tests should be there or here. I think this is fine 😄

talldan reviewed (commented):
I think this is working really well.

There are a couple of minor bugs that I've spotted - the remove and duplicate actions in the block settings menu in List View don't seem to work with a multi-selection (though Group seems to work fine). They only duplicate or remove the first selected block.

The 'Remove block' label doesn't seem to take into account that multiple blocks are selected either, so it might be that only one selected block is passed into the component.

I think if the settings menu within a multi-selection is used, the action can probably apply to multiple blocks. If a settings menu from outside of a multi-selection is used it should maybe apply to only the focused block.

Now that the keyboard navigation feature is working, it might be good to get @alexstine's feedback. Alex - you should now be able to test multi-selecting blocks in ListView on this PR by holding down the shift key and using up/down arrow keys.

In my testing with a screenreader I hear '1 block selected', '2 blocks selected' while I'm selecting blocks—I think that matches how multi-selection works in the main canvas. If I release shift and use arrow keys to move focus back over the blocks I can hear a message like 'Paragraph (selected block)' for selected blocks.

andrewserong commented:
Thanks again for the detailed testing @talldan! I'll dig more into your feedback and notes tomorrow and give the PR a rebase to re-test against other changes that have landed in the last couple of weeks.

andrewserong committed:
Update initialPosition param to be experimental

andrewserong committed:
Update TreeGrid readme with documentation for the three callback functions

andrewserong committed:
Add changelog entry

alexstine commented:
@andrewserong Accessibility with the keyboard is getting much better, great work. 👍 My only comment is about the unselect behavior. If I select a Paragraph Block and a Table Block, then unselect the Table Block, it announces "1 Block selected, Paragraph Block selected". Is there anyway to change this so it says which item was "unselected"? E.g. "1 Blocks selected, Table Block unselected". It may even be a good idea to change it around so name of Block comes first followed by how count.

Thoughts?

Thanks.

andrewserong committed:
Pass in selectedClientIds to dropdown

andrewserong committed:
Prevent shift+click on expand toggle from opening a new window

andrewserong committed:
Try announcing deselected blocks

andrewserong commented:
@talldan I've updated this PR to pass in the `selectedClientIds` to the dropdown menu so that when there's a multiple selection, the dropdown menu applies to the whole selection. If a user clicks on one of the blocks that isn't part of the selection, then much like the grad behaviour, it only applies to the clicked on block. I _think_ this should now fix most of the issues you raised.

@javierarce great catch about the expand / collapse button. I've added in an `event.preventDefault` call and that appears to be fixed now (toggling the visibility has no effect on the selection).

andrewserong commented:
@alexstine, thanks for giving this another test! 🙇 That's a great idea about announcing the blocks that are deselected. I've made an attempt at adding in an announcement, but I'm not sure if it's quite right. It now announces the name of the block that has been deselected if there is only one. If there is greater than one deselected block, then it announces the number of blocks deselected.

Because the code that announces the selection of blocks is spread across different areas, I'm not sure how feasible it is in this PR to move the logic all to the one place, which means that we might be limited in terms of the order in which things are announced, unless there's another way to manipulate it. It now reads out both the block that is selected and the one that is deselected, so I could imagine that being confusing. I think the block that is selected is announced via the component we are currently focused on, rather than a call to `speak`, so it might need some more tweaking to get it right.

Please let me know if you have further feedback on how it should work, and I'm happy to keep iterating!

talldan commented on packages/components/src/tree-grid/README.md:
```diff
@@ -108,10 +108,33 @@ function TreeMenu() {
+
+-   Type: `Function`
+-   Required: No
+
```
Thanks for documenting the expand/collapse props too!

talldan commented on packages/block-editor/src/components/list-view/use-block-selection.js:
```diff
@@ -97,10 +108,44 @@ export default function useBlockSelection() {
+			}
+
+			if ( label ) {
+				speak( label );
```
This message works pretty well in testing.

In terms of code, it could be an option to try moving this announcement to the `multiSelect` action which already has a call to `speak`.

It should hopefully be possible to combine it with the existing message. Something like '2 blocks selected (Paragraph deselected)'. I'll defer to Alex's call on what the best text would be.

I think it'd be ok to try this out in a separate PR.

andrewserong replied:
Thanks for confirming, Dan! I agree, this PR has already gotten fairly big, so it'd probably be a good thing to explore separately as a follow-up.

talldan reviewed (approved):
This is working really well. Thanks for all the hard work on this PR!

andrewserong commented:
Thanks for reviewing again @talldan! I'll leave this PR overnight just in case anyone else has feedback they'd like to see incorporated before merging. If there aren't any objections, I'll then merge it in tomorrow. Thanks again for all the reviews, everyone 🙇

alexstine reviewed (approved):
This looks good for accessibility. Can improve in separate PR where diffs will be smaller and easier to test.

Thanks for your hard work on this @andrewserong . Certainly looks like a ton more code involved for the keyboard part of this.

ciampo commented on packages/components/src/tree-grid/README.md:
```diff
@@ -108,10 +108,33 @@ function TreeMenu() {
 
 `TreeGrid` should always have children.
 
+###### onChangeRow( event: Event, startRow: HTMLElement, destinationRow: HTMLElement )
```
Usually `onChange` is associated more to a value change for a input-like element — maybe `onFocusRow` is a better name?

andrewserong replied:
Oh, great idea, that's a better fit. Updated in https://github.com/WordPress/gutenberg/pull/38314/commits/5511896de7cc60499a24ffece4cefb64c68b1b15

ciampo commented on packages/components/CHANGELOG.md:
```diff
@@ -22,6 +22,7 @@
 -  `TreeGrid` accessibility: improve browser support for Left Arrow focus to parent row in child row.  ([#38639](https://github.com/WordPress/gutenberg/pull/38639))
 -  `TreeGrid` accessibility: Add Home/End keys for better keyboard navigation. ([#38679](https://github.com/WordPress/gutenberg/pull/38679))
 -   Add `resolvePoint` prop to `FocalPointPicker` to allow updating the value of the picker after a user interaction ([#38247](https://github.com/WordPress/gutenberg/pull/38247))
+-   Add `onChangeRow` callback to the `TreeGrid` component, fired when focus is shifted from one row to another via Up and Down arrow keys. ([#38314](https://github.com/WordPress/gutenberg/pull/38314))
```
Should we also mention that the change in behaviour of `TreeGrid` related to the SHIFT key?

andrewserong replied:
Good idea, I've updated the changelog entry to cover that in https://github.com/WordPress/gutenberg/pull/38314/commits/5511896de7cc60499a24ffece4cefb64c68b1b15

ciampo reviewed (commented):
> I'll leave this PR overnight just in case anyone else has feedback they'd like to see incorporated before merging. If there aren't any objections, I'll then merge it in tomorrow. Thanks again for all the reviews, everyone 🙇

Sorry for my late review, @andrewserong ! Unfortunately I'm lagging a bit behind with PR reviews these days 😅  I only left a couple of tiny inline comments, nothing that should really delay this PR further.

Thank you!

andrewserong committed:
Rename onChangeRow to onFocusRow, update changelog entry

andrewserong commented:
Alrighty, tests are passing, so I'll merge this in now. Thanks so much again @noisysocks, @gwwar, @talldan, @alexstine, @ciampo, and @javierarce for all the reviews, testing, and feedback — it's been immeasurably helpful in developing this feature (and I've learnt a lot in the process, too). Kudos! 🙇

jasmussen commented:
🔥

andrewserong commented:
I've just opened a PR in #38942 as a follow-up to add in some unit tests for the `TreeGrid` component's callback functions — this is mostly to ensure that any future changes to that component don't affect the expected behaviour in the list view.

ellatrix commented on packages/block-editor/src/store/actions.js:
```diff
@@ -259,10 +259,15 @@ export function stopMultiSelect() {
+export const multiSelect = (
+	start,
+	end,
+	__experimentalInitialPosition = 0
```
@andrewserong Why did we add this parameter? How does `initialPosition` make any sense for a multi block selection? `initialPosition` implies collapsed selection, while multi selection implies an uncollapsed selection.

andrewserong replied:
@ellatrix the existing `initialPosition` state is used in the writing flow to control whether or not to shift focus to the editor (for single block selection), so I thought it might make sense to re-use that same logic for multiple block selection so that the logic is as consistent as we can make it. In this case, we're using it effectively like a boolean, so setting to null says "there is no initial position" (and therefore don't focus in the editor canvas), and with the default `0` value, the position is "within the editor canvas".

For accessibility it was important that we don't shift focus to the editor canvas when multi-selecting from the list view. If there's a more semantic way to do this than using `initialPosition` we can definitely refactor it (one of the reasons I think @talldan suggested we use the `__experimental` prefix for the moment).

Do you have any guidance on how best to preserve the behaviour of skipping focusing on the editor canvas, if we were to refactor this?

ellatrix replied:
Yes, I understood the intent after digging a bit further. :) The naming is a bit confusing and I'm not sure if reusing the `0` value makes sense to mean selecting all content. Another thing to keep in mind is that focus and selection are two separate things. Even if focus has to stay in the List View, we still need to select the block contents. Anyway, it's not important right now :)

andrewserong replied:
Okay, great. Thanks for confirming @ellatrix, please do ping me if it does become important / needs updating, though, and I can follow up!

andrewserong commented:
I've opened up another small follow-up PR in https://github.com/WordPress/gutenberg/pull/39272 to look at adding support for shift-selecting via the Home and End keys.
